test/jit/test_freezing.py

```diff
@@ -1585,6 +1585,124 @@ def forward(self, x):
+                self.b1 = torch.rand([w1_dim])
+                self.w2 = torch.rand([w2_dim, 5])
+                self.b2 = torch.rand([w2_dim])
+
+            def forward(self, in_tensor1):
+                res1 = torch._C._nn.linear(in_tensor1, self.w1, self.b1)
+                res3 = torch._C._nn.linear(res1, self.w2, self.b2)
+                res2 = torch._C._nn.linear(in_tensor1, self.w2, self.b2)
+                res4 = torch._C._nn.linear(res1, self.w1, self.b1)
+                return res2, res3, res4
+
+        mod_eager = ModMultLinear().eval()
+        script_mod = torch.jit.script(mod_eager)
+        op_graph = script_mod.graph
+        print(op_graph)
+
+        FileCheck().check_count("aten::linear", 4, exactly=True).run(op_graph)
+        # successively no-ops with non-const inputs
+        self.run_pass("concat_frozen_linear", op_graph)
+        FileCheck().check_count("aten::linear", 4, exactly=True).run(op_graph)
+
+        script_mod = torch.jit.freeze(script_mod)
+        op_graph = script_mod.graph
+        self.run_pass("concat_frozen_linear", op_graph)
+        FileCheck().check_count("aten::linear", 2, exactly=True).run(op_graph)
+
+        test_val1 = torch.rand([50, 5])
+        self.assertEqual(mod_eager.forward(test_val1), script_mod(test_val1))
+
+    def test_linear_concat_different_input(self):
+        """
+        There should be no change to the graph due to the optimization pass
+        due to the two input tensors being different
+        """
+
+        # Freezing requires that the graph be a module
+        class ModMultLinear(nn.Module):
+            def __init__(self, w1_dim, w2_dim):
+                super(ModMultLinear, self).__init__()
+                self.w1 = torch.rand([w1_dim, 5])
+                self.b1 = torch.rand([w1_dim])
+                self.w2 = torch.rand([w2_dim, 5])
+                self.b2 = torch.rand([w2_dim])
+
+            def forward(self, in_tensor1, in_tensor2):
+                res1 = torch._C._nn.linear(in_tensor1, self.w1, self.b1)
+                res2 = torch._C._nn.linear(in_tensor2, self.w2, self.b2)
+                return res1, res2
+
+        mod_eager = ModMultLinear(5, 5).eval()
+        script_mod = torch.jit.script(mod_eager)
+        op_graph = script_mod.graph
+        print(op_graph)
+
+        FileCheck().check_count("aten::linear", 2, exactly=True).run(op_graph)
+        # successively no-ops with non-const inputs
+        self.run_pass("concat_frozen_linear", op_graph)
+        FileCheck().check_count("aten::linear", 2, exactly=True).run(op_graph)
+
+        script_mod = torch.jit.freeze(script_mod)
+        op_graph = script_mod.graph
+        self.run_pass("concat_frozen_linear", op_graph)
+        FileCheck().check_count("aten::linear", 2, exactly=True).run(op_graph)
+
+        test_val1 = torch.rand([50, 5])
+        test_val2 = torch.rand([50, 5])
+        self.assertEqual(mod_eager(test_val1, test_val2), script_mod(test_val1, test_val2))
```

Comment:
I would factor out this block to take in a module, tuple of inputs, and number of linear nodes before/after calling the pass, since we are using this block 3 times.